### ISSUE TITLE
Performance seriously worse after updating to 2.2.0.5 from 2.2.0.4

### ISSUE THREAD

**Issue by Ben305** [docker/for-mac#4423]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 430E49E5-D565-419C-B0D0-3C8304281DF0/20200404081442

I'm using DDEV to run the dev environment for Pimcore (PHP/Symfony based web app). After upgrading Docker for 2.2.0.4 to 2.2.0.5, the performance became seriously worse. Requests sent to the Pimcore webserver (Apache with php-fpm) time out, the system is unusable on a MacBook Pro from 2016/2019.

### Expected behavior

The dev environment should be as useable as with Docker 2.2.0.4

### Actual behavior

The dev environment isn't useable with Docker 2.2.0.5. Switching back to 2.2.0.4 fully resolves the problem,.

### Information

  - macOS Version: 10.15.4
  - Happens on multiple devices. MacBook Pro from 2016, MacBook Pro from 2019
  - Switchting back to 2.2.0.4 fully resolves the problem

### Diagnostic logs

```
Docker for Mac: version...

```

### Steps to reproduce the behavior

  1. Use dev environment with Docker 2.2.0.4 -> everything is fine
  2. Use dev environment with Docker 2.2.0.5 -> system is unusable

**Comment by you-think-you-are-special** [comment]
The same issue. Docker 2.2.0.5 uses 100% CPU like it was in 2.2.0.4. But it also crashes all the time. Diagnostic not working.

**Comment by AlexBachmann** [comment]
Confirmed. osxfx is slow as hell after the update. Needed to revert.

**Comment by ceaglest** [comment]
Thanks for filing an issue! I was about to update but will stick with 2.2.0.4 until this is resolved.

**Comment by rimbu002** [comment]
Same here, it's horribly slow.
What's the easiest way to downgrade to the previous version?

**Comment by Ben305** [comment]
@rimbu002 Download version 2.2.0.4 from here: https://docs.docker.com/docker-for-mac/release-notes/

**Comment by ulrich** [comment]
Hello,

We are investigating about this trouble for finding a way to reproduce the issue.

**Actually we cannot reproduce the performance issue with the following stack profile :**

- a docker compose stack with following images : redis:alpine, mariadb:10.4 and pimcore:PHP7.2-apache
- desktop side, two shared folders for the pimcore data (mariadb) and pimcore front (apache)
- Not ddev environment set.

**The docker-compose file used for testing pimcore :**

`
version: '2.0'
services:
  redis:
    image: redis:alpine
    container_name: pimcore-redis

  db:
    image: mariadb:10.4
    container_name: pimcore-mariadb
    working_dir: /application
    command: [mysqld, --character-set-server=utf8mb4, --collation-server=utf8mb4_unicode_ci, --innodb-file-format=Barracuda, --innodb-large-prefix=1, --innodb-file-per-table=1]
    volumes:
      - /Users/ulrich/projects/pimcore-docker-compose-2/data:/var/lib/mysql
    environment:
      - MYSQL_ROOT_PASSWORD=ROOT
      - MYSQL_DATABASE=pimcore
      - MYSQL_USER=pimcore
      - MYSQL_PASSWORD=pimcore

  php:
    image: dpfaffenbauer/pimcore:PHP7.2-apache
    container_name: pimcore-php
    volumes:
      - /Users/ulrich/projects/pimcore-docker-compose-2/apache:/var/www/html:cached
    ports:
     - "2000:80"
     - "2001:443"
    links:
     - db:db
     - redis:redis
`

**The apache bench results ran on several pages nearly return the same values between 2.2.0.4 and 2.2.0.5 (example for one page) :**

- (2.2.0.4) ab -n 100 http://localhost:2000/en/shop/Products\~c564
Document Length: 153615 bytes
Time taken for tests: 71.925 seconds
Time per request: 719.246 [ms] (mean)

- (2.2.0.5) ab -n 100 http://localhost:2000/en/shop/Products\~c564
Document Length: 153615 bytes
Time taken for tests: 67.520 seconds
Time per request: 675.199 [ms] (mean)

**Comment by ulrich** [comment]
@Ben305 and all. Can you give use a little more information for helping to reproduce and debugging the issue please? Thank you for your feedback.

**Comment by AlexBachmann** [comment]
My first  assumption was, that it has something to do with the volume binding. So this is the setup I use:

´´´
volumes:
      - type: bind
        source: .
        target: /var/www/html
        consistency: delegated
´´´

**Comment by Ben305** [comment]
This is an equal snippet from my ddev environment:

```
volumes:
  - type: bind
    source: ../
    target: /var/www/html
        
    consistency: cached
```

I'm also running a php application based on symphony running on the Apache server.

**Comment by stephen-turner** [comment]
Sorry to say, we have been trying but have failed to reproduce any significant performance difference between 2.2.0.4 and 2.2.0.5. Does anyone have more detailed reproduction steps?

**Comment by Ben305** [comment]
@stephen-turner I just checked Docker 2.2.0.4 and 2.2.0.5 using the Pimcore official demo using the DDEV environment. I can easily reproduce the performance issues:

Apache Bench using 2.2.0.4:
```
ab -n 20 http://127.0.0.1:8000/
This is ApacheBench, Version 2.3 <$Revision: 1843412 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 127.0.0.1 (be patient).....done


Server Software:        nginx/1.17.9
Server Hostname:        127.0.0.1
Server Port:            8000

Document Path:          /
Document Length:        94562 bytes

Concurrency Level:      1
Time taken for tests:   19.833 seconds
Complete requests:      20
Failed requests:        0
Total transferred:      1915120 bytes
HTML transferred:       1891240 bytes
Requests per second:    1.01 [#/sec] (mean)
Time per request:       991.649 [ms] (mean)
Time per request:       991.649 [ms] (mean, across all concurrent requests)
Transfer rate:          94.30 [Kbytes/sec] received
```

Apache Bench using 2.2.0.5:
```
ab -n 20 http://127.0.0.1:8000/
This is ApacheBench, Version 2.3 <$Revision: 1843412 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 127.0.0.1 (be patient).....done


Server Software:        nginx/1.17.9
Server Hostname:        127.0.0.1
Server Port:            8000

Document Path:          /
Document Length:        94562 bytes

Concurrency Level:      1
Time taken for tests:   109.041 seconds
Complete requests:      20
Failed requests:        0
Total transferred:      1915120 bytes
HTML transferred:       1891240 bytes
Requests per second:    0.18 [#/sec] (mean)
Time per request:       5452.057 [ms] (mean)
Time per request:       5452.057 [ms] (mean, across all concurrent requests)
Transfer rate:          17.15 [Kbytes/sec] received
```

I just pushed the Pimcore Demo including the DDEV config: 
https://github.com/Ben305/pimcore-docker-perf-sample

Assuming that you have already DDEV installed, you just need to execute these commands:
```bash
ddev start
ddev composer install 
ddev exec ./vendor/bin/pimcore-install
```

**Comment by stephen-turner** [comment]
Thanks @Ben305, we'll have a look at that example.

**Comment by stephen-turner** [comment]
@Ben305 Thanks again for your help with this. We have reproduced it with your example. We're still trying to understand why, and what is different about your workload compared with the others that we tried.

**Comment by Ben305** [comment]
@stephen-turner Glad to hear that. Let me know when a fix is available, I'd like to verify that, too.

**Comment by ulrich** [comment]
@Ben305 a workaround could be to use an overriding configuration file of the docker-compose provided by DDEV (**.ddev/.ddev-docker-compose-base.yaml**). This file must corresponding with the target service (eg. : **docker-compose.web.yaml**). So, you can manage the issue like that and use the short notation as describe in the Docker Compose document page (https://docs.docker.com/compose/compose-file/) :

Example:
`version: '3.6'

services:
  web:
      volumes:      
        - ../:/var/www/html:cached
`

Before the overriding configuration:
`➜  pimcore-docker-perf-sample-2.2.0.5 git:(master) ✗ /Applications/Docker.app/Contents/MacOS/com.docker.osxfs state
Exported directories:
 - /Users to /Users (nodes_/Users table size: 33)
 - /Volumes to /Volumes (nodes_/Volumes table size: 0)
 - /private to /private (nodes_/private table size: 0)
 - /tmp to /tmp (nodes_/private/tmp table size: 40)
Container-mounted directories:
 - /Users/ulrich/projects/pimcore/pimcore-docker-perf-sample-2.2.0.5 into docker/80c26b32162fdf5670eda215df0cccc9668e37f048c10b3da36643254246c7be (state=default)
 - /Users/ulrich/projects/pimcore/pimcore-docker-perf-sample-2.2.0.5/.ddev into docker/87dc007c75ffad67df4cb45c75f5ff4a2dc79d587542843c5d5f22c90dd7123d (state=default)
 - /Users/ulrich/projects/pimcore/pimcore-docker-perf-sample-2.2.0.5/.ddev into docker/80c26b32162fdf5670eda215df0cccc9668e37f048c10b3da36643254246c7be (state=default)
`

After the overriding configuration:
`➜  pimcore-docker-perf-sample-2.2.0.5 git:(master) ✗ /Applications/Docker.app/Contents/MacOS/com.docker.osxfs state
Exported directories:
 - /Users to /Users (nodes_/Users table size: 4733)
 - /Volumes to /Volumes (nodes_/Volumes table size: 0)
 - /private to /private (nodes_/private table size: 0)
 - /tmp to /tmp (nodes_/private/tmp table size: 65)
Container-mounted directories:
 - /Users/ulrich/projects/pimcore/pimcore-docker-perf-sample-2.2.0.5 into docker/754cb84ab4b79d008f4c307f2bddab2ee90895176a247163e7ef37147d557dc6 (state=**cached**)
 - /Users/ulrich/projects/pimcore/pimcore-docker-perf-sample-2.2.0.5/.ddev into docker/d98e0ca13b9e02521742a06504b0e38e0ce7775fe637cdc242c3a3c4835c065c (state=default)
 - /Users/ulrich/projects/pimcore/pimcore-docker-perf-sample-2.2.0.5/.ddev into docker/754cb84ab4b79d008f4c307f2bddab2ee90895176a247163e7ef37147d557dc6 (state=default)
`

**Comment by djs55** [comment]
@Ben305 thanks for your patience on this one. I have a candidate fix in a build if you'd like to give it a go: https://download-stage.docker.com/mac/edge/45075/Docker.dmg . If you get a chance, let me know if that helps or not. If not, could you upload a fresh set of diagnostics and quote the id here? Thanks in advance!

**Comment by Ben305** [comment]
@djs55 Seems quite a good candidate for a fix, I'm seeing equal or even better values using that edge version 👍

**Comment by dietervdw-spotify** [comment]
I was wondering why my laptops fan was on full blast the whole day, even when it was just sitting there idle. Turns out Docker 2.2.0.5 is burning 140% CPU at all times, even though there are no containers running.

![image](https://user-images.githubusercontent.com/36199260/81327466-973e6200-909b-11ea-8a50-2d5e81355006.png)

Is that related to this issue?

**Comment by stephen-turner** [comment]
No, it's unrelated. There are other tickets about it though.

**Comment by stephen-turner** [comment]
Fixed in 2.3, closing this ticket.

**Comment by petr-ujezdsky** [comment]
You could also try to lower the docker daemon's CPU core count from `max` to `max - 1`.
I know it sounds strange, but the `max` (default) setting actually makes IO performance much worse.

For details and example see my ticket #3771.
